worlds/blasphemous/Rules.py

```diff
@@ -1,1479 +1,3333 @@
+                    self.has("D09Z01S03[N]", player))
+        elif boss == "Rooftops":
+            return has_boss_strength("crisanta") and \
+                (self.has("D06Z01S25[W]", player) or \
+                    self.has("D06Z01S25[E]", player))
+        elif boss == "Ossuary":
+            return has_boss_strength("isidora") and \
+                self.has("D01BZ08S01[W]", player)
+        elif boss == "Mourning":
+            return has_boss_strength("sierpes") and \
+                self.has("D20Z02S08[E]", player)
+        elif boss == "Graveyard":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D02Z03S23[E]", player) and \
+                        self.has("D02Z02S14[W]", player) and \
+                            self.has("Wall Climb Ability", player)
+        elif boss == "Jondo":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    (self.has("D20Z01S05[W]", player) or \
+                        self.has("D20Z01S05[E]", player)) and \
+                            (self.has("D03Z01S03[W]", player) or \
+                                self.has("D03Z01S03[SW]", player))
+        elif boss == "Patio":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D06Z01S18[E]", player) and \
+                        (self.has("D04Z01S04[W]", player) or \
+                            self.has("D04Z01S04[E]", player)) or \
+                                self.has("D04Z01S04[Cherubs]", player)
+        elif boss == "Wall":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D09BZ01S01[Cell24]", player) and \
+                        (self.has("D09Z01S01[W]", player) or \
+                            self.has("D09Z01S01[E]", player))
+        elif boss == "Hall":
+            return has_boss_strength("laudes") and \
+                (self.has("D08Z03S03[W]", player) or \
+                    self.has("D08Z03S03[E]", player))
+        elif boss == "Perpetua":
+            return has_boss_strength("perpetua")
+        elif boss == "Legionary":
+            return has_boss_strength("legionary")
+        
+    def _blasphemous_guilt_rooms(self, player, number: int):
+        total: int = 0
+
+        if self.has("D01Z04S17[W]", player):
+            total += 1
+        if self.has("D02Z02S06[E]", player):
+            total += 1
+        if self.has("D03Z03S14[W]", player):
+            total += 1
+        if self.has("D04Z02S17[W]", player):
+            total += 1
+        if self.has("D05Z01S17[W]", player):
+            total += 1
+        if self.has("D09Z01S13[E]", player):
+            total += 1
+        if self.has("D17Z01S12[E]", player):
+            total += 1
```

Comment:
Could define the list guilt_rooms = ["D01Z04S17[W]", ....] somewhere and then do

`total = sum(1 for item in guilt_rooms if self.has(item, player))`

Or, if you wanna get wacky with it (which I always do)

`total = sum(int(self.has(item, player)) for item in guilt_rooms)`

I think the "int()" there might even be unnecessary

Reply:
A lot of the following code has the same potential improvements as I've been laying out, so I won't repeat myself.

It might be a matter of taste, but I usually prefer having data structures, and then have the actual code be very short by iterating / doing a lookup to those data structures. It's often faster, too (especially in the case of dict lookups instead of if-elif chains)

Reply:
If it's faster then I'll do it